### ISSUE TITLE
Items in the references list added from connectors/plugins do not show up

### ISSUE THREAD

**Issue by rbeckman-nextgen** [nextgenhealthcare/connect#3211]
We allow connectors and code template plugins to inject their own templates. However it becomes a chicken-and-egg scenario, because the reference list factory is called in the constructor of components like MirthSyntaxTextArea, within the constructor of a connector settings panel. So at the time setup() is called, not all plugins have yet been loaded.

Imported Issue. Original Details:
Jira Issue Key: MIRTH-3305
Reporter: narupley
Created: 2014-05-29T10:43:00.000-0700

**Comment by rbeckman-nextgen** [comment]
In order to guarantee that code templates added from connectors/plugins show up correctly all the time, we need to be able to guarantee the order by which the extension is loaded (MIRTH-3034). Plugins should also no longer add items to the reference list directly. Instead they should implement a CodeTemplatePlugin class. For connectors, we can add a new codeTemplateClassName if a connector needs to add reference list items. Then we need to ensure that CodeTemplatePlugin classes are loaded first. 

Imported Comment. Original Details:
Author: wayneh
Created: 2014-05-30T11:11:52.000-0700

**Comment by rbeckman-nextgen** [comment]
Revision 7276: Changed how code templates get added from extensions. Instead of appending to the references list on start, the data type plugins (and all other extensions) now implement a CodeTemplatePlugin class, which is instantiated and added in LoadedExtensions before anything else. Connectors also use this method, so the File/HTTP reference items code has been moved to separate classes.

Imported Comment. Original Details:
Author: narupley
Created: 2014-08-11T15:47:09.000-0700

**Comment by rbeckman-nextgen** [comment]
Verified in 3.0.3 that some of the code templates do not show up in the references list. I also downloaded and installed several extensions and data types to see whether or not they showed up as well. In 3.1, various code templates that were not readily available in 3.0.3 now display properly in the newest build. The extensions that I tested with were seemingly unaffected as they showed up in both builds.

Imported Comment. Original Details:
Author: jaysenp
Created: 2014-09-04T11:13:05.000-0700
